File: Ch1-Introduction/Merge-Sort/Haskell/MergeSort.hs
module MergeSort (mergesort) where

mergesort :: [Int] -> [Int]
mergesort list =
  case list
  of []        -> []
     (_:rest)
      | null rest -> list
      | otherwise -> conquer (mergesort left) $ mergesort right
        where (left, right) = divide list

-- private functinons
divide :: Ord a => [a] -> ([a], [a])
divide xs = splitAt (length xs `div` 2) xs

conquer :: Ord a => [a] -> [a] -> [a]
conquer xs ys =
  case (xs, ys) of
    ([], _) -> ys
    (_, []) -> xs
    (x0:xTail, y0:yTail)
      | x0 < y0   -> x0:conquer xTail ys
      | otherwise -> y0:conquer xs yTail

File: Ch1-Introduction/Merge-Sort/Python/merge_sort.py
def mergesort(xs):
    if len(xs) < 2:
        return xs
    left_xs, right_xs = _divide(xs)
    return _conquer(mergesort(left_xs), mergesort(right_xs))

def _divide(xs):
    return split_at(xs, len(xs) / 2)

# not thread safe
def _conquer(left_xs, right_xs):
    # copy the lists to avoid side-effects
    left = list(left_xs)
    right = list(right_xs)

    sorted_xs = []
    while left and right:
        left0 = head(left)
        right0 = head(right)
        if left0 < right0:
            sorted_xs.append(left0)
            left = tail(left)
        else:
            sorted_xs.append(right0)
            right = tail(right)
    sorted_xs += right + left
    return sorted_xs

# helper methods that could probably be pulled into a library
def split_at(xs, index):
    return (xs[:index], xs[index:])

def tail(xs):
    return xs[1:]

def head(xs):
    return xs[0]

